packages/http/src/merge-patch.ts

```diff
@@ -0,0 +1,465 @@
+            // perform the mergePatch transform of the ModelProperty
+            if (
+              overrides?.erasable !== false &&
+              (prop.optional || prop.defaultValue !== undefined)
+            ) {
+              clone.type = nullable(realm, clone.type);
+            }
+            clone.optional =
+              overrides?.optional ?? (isDiscriminatedProperty(program, prop) ? false : true);
+            clone.defaultValue = undefined;
+          }
+
+          ctx.program.stateMap(HttpStateKeys.mergePatchProperty).set(clone, prop);
+        },
+      },
+    };
+
+    const self: Mutator = {
+      name: `MergePatch${visibilityMode}`,
+      Union: {
+        filter: () => MutatorFlow.DoNotRecur,
+        mutate: (union, clone, program) => {
+          for (const [key, member] of union.variants) {
+            overrideDiscriminatedUnionProperty(program, member);
+            if (isMergePatchSubject(member.type)) {
+              const variant: UnionVariant = {
+                ...member,
+                type: cachedMutateSubgraph(program, _optionalInteriorMutator ?? self, member.type)
+                  .type,
+              };
+              clone.variants.set(key, variant);
+            }
+          }
+
+          rename(ctx.program, clone, nameTemplate);
+        },
+      },
+      Model: {
+        filter: () => MutatorFlow.DoNotRecur,
+        mutate: (model, clone, program, realm) => {
+          if ($(realm).array.is(model) && isMergePatchSubject(model.indexer!.value)) {
+            clone.indexer = {
+              key: model.indexer!.key,
+              value: cachedMutateSubgraph(
+                program,
+                _replaceInteriorMutator ?? self,
+                model.indexer!.value,
+              ).type,
+            };
+          } else if ($(realm).record.is(model) && isMergePatchSubject(model.indexer!.value)) {
+            clone.indexer = {
+              key: model.indexer!.key,
+              value: mutateSubgraph(
+                program,
+                [_optionalInteriorMutator ?? self], // records are always CreateOrUpdate
+                model.indexer!.value,
+              ).type,
+            };
+          }
+
+          for (const [key, prop] of model.properties) {
+            if (!isVisible(program, prop, visibilityFilter)) {
+              // Property is not visible, remove it
+              const clonedProp = clone.properties.get(key);
+              if (clonedProp) {
+                clone.properties.delete(key);
+                realm.remove(clonedProp);
+              }
+            } else if (!isMetadata(program, prop)) {
+              const mutated = mutateSubgraph(program, [mpMutator], prop);
+              const mutatedProp = mutated.type as ModelProperty;
+              mutatedProp.model = clone;
+              clone.properties.set(key, mutatedProp);
+            } else {
+              const decorator: string | undefined = isPathParam(program, prop)
```

Comment:
Feel like this if piping could use some common `getHttpPropertyKind` helper by extracting some logic in `http-property` but probably can do that later as it seems there is a few other places to clean up